File: ui/goth/primitives/combobox.go
package primitives

import "github.com/a-h/templ"

// Combobox (P50, family F4). A text input paired with a listbox of options the
// SERVER owns. Compound parts the caller composes: Combobox (root, x-data=
// gothCombobox), ComboboxInput (the editable role=combobox field), ComboboxListbox
// (the role=listbox popup), ComboboxOption (a role=option submit control), and
// ComboboxEmpty (the no-results message).
//
// Server ownership. The server owns the option DATA and, in server-filter mode,
// the FILTERING and empty-state markup: typing re-fetches a fresh option list
// (async option replacement). In the default client-filter mode the controller
// hides non-matching already-rendered options and toggles ComboboxEmpty; either
// way the server owns the authoritative value.
//
// No-JS baseline. Each ComboboxOption is a native <button type="submit"> carrying
// name/value, so with the listbox rendered server-open (ComboboxProps.Open) a user
// picks an option and the enclosing form submits the value with no JavaScript —
// the same round-trip an HTMX-enhanced option performs via hx-* on Base.Attributes.
// The input carries its own Name so a free-text query also round-trips (the
// "native input + server round-trip" baseline).
//
// Enhancement. gothCombobox opens the listbox on focus/typing, moves an active
// option with the arrow keys via aria-activedescendant (focus stays on the input),
// activates the active option on Enter, and dismisses on Escape/outside press
// through the shared overlay mechanics — dispatching goth:open/goth:close/
// goth:select. data-slot hooks: combobox, input, listbox, option, empty.

// ComboboxFilter selects who owns filtering. The zero value is client filtering.
type ComboboxFilter string

const (
	// ComboboxFilterClient hides non-matching already-rendered options in the
	// browser (the zero value).
	ComboboxFilterClient ComboboxFilter = "client"
	// ComboboxFilterServer leaves filtering and the empty state to the server
	// (async option replacement); the controller only opens the listbox and moves
	// the active option.
	ComboboxFilterServer ComboboxFilter = "server"
)

// Valid reports whether f is a known filter mode.
func (f ComboboxFilter) Valid() bool {
	return f == ComboboxFilterClient || f == ComboboxFilterServer
}

func (f ComboboxFilter) attr() string {
	if f == ComboboxFilterServer {
		return "server"
	}
	return "client"
}

// ComboboxProps configures the Combobox root.
type ComboboxProps struct {
	Base
	// Filter selects client (default) or server/async filtering.
	Filter ComboboxFilter
	// Open renders the listbox server-open (the readable no-JS baseline). Zero
	// value is closed (only the input shows without JavaScript).
	Open bool
}

// ComboboxInputProps configures the editable combobox field.
type ComboboxInputProps struct {
	Base
	// Name is the submitted free-text query field (the no-JS round-trip carries it).
	Name string
	// Value is the current input text.
	Value string
	// Placeholder is the empty-field hint.
	Placeholder string
	// Listbox is the ComboboxListbox id this input controls (aria-controls).
	Listbox string
	// Required marks the field required.
	Required bool
	// Disabled disables the field.
	Disabled bool
	// Invalid marks the field invalid (aria-invalid + data-invalid).
	Invalid bool
}

// ComboboxListboxProps configures the role=listbox popup.
type ComboboxListboxProps struct {
	Base
	// Open must match ComboboxProps.Open so the popup's data-state agrees with the
	// root for the server-owned no-JS baseline.
	Open bool
	// Label is the listbox accessible name (aria-label).
	Label string
}

// ComboboxOptionProps configures one role=option submit control.
type ComboboxOptionProps struct {
	Base
	// Name is the submitted value field name; Value is the submitted value.
	Name  string
	Value string
	// Selected marks the committed option (aria-selected + data-selected).
	Selected bool
	// Disabled marks the option non-actionable.
	Disabled bool
}

// ComboboxEmptyProps configures the no-results message.
type ComboboxEmptyProps struct {
	Base
	// Open renders the empty state visible (server-owned no-results state).
	Open bool
}

func comboboxAttrs(p ComboboxProps) templ.Attributes {
	return ownedAttrs(p.Base, templ.Attributes{
		"data-slot":   "combobox",
		"data-state":  openState(p.Open),
		"data-filter": p.Filter.attr(),
		"x-data":      "gothCombobox",
	})
}

func comboboxInputAttrs(p ComboboxInputProps) templ.Attributes {
	owned := templ.Attributes{
		"data-slot":         "input",
		"type":              "text",
		"role":              "combobox",
		"aria-autocomplete": "list",
		"aria-expanded":     "false",
		"autocomplete":      "off",
		"x-on:input":        "onInput($event)",
		"x-on:focus":        "onFocus($event)",
		"x-on:keydown":      "onKeydown($event)",
	}
	if p.Name != "" {
		owned["name"] = p.Name
	}
	if p.Value != "" {
		owned["value"] = p.Value
	}
	if p.Placeholder != "" {
		owned["placeholder"] = p.Placeholder
	}
	if p.Listbox != "" {
		owned["aria-controls"] = p.Listbox
	}
	if p.Required {
		owned["required"] = true
	}
	if p.Disabled {
		owned["disabled"] = true
	}
	if p.Invalid {
		owned["aria-invalid"] = "true"
		owned["data-invalid"] = "true"
	}
	return ownedAttrs(p.Base, owned)
}

func comboboxListboxAttrs(p ComboboxListboxProps) templ.Attributes {
	owned := templ.Attributes{
		"data-slot":  "listbox",
		"role":       "listbox",
		"data-state": openState(p.Open),
	}
	if p.Label != "" {
		owned["aria-label"] = p.Label
	}
	return ownedAttrs(p.Base, owned)
}

func comboboxOptionAttrs(p ComboboxOptionProps) templ.Attributes {
	owned := templ.Attributes{
		"data-slot":     "option",
		"role":          "option",
		"type":          "submit",
		"aria-selected": boolString(p.Selected),
		"x-on:click":    "select($event)",
	}
	if p.Value != "" {
		owned["data-value"] = p.Value
		owned["value"] = p.Value
	}
	if p.Name != "" {
		owned["name"] = p.Name
	}
	if p.Selected {
		owned["data-selected"] = "true"
	}
	if p.Disabled {
		owned["disabled"] = true
		owned["aria-disabled"] = "true"
	}
	return ownedAttrs(p.Base, owned)
}

func comboboxEmptyAttrs(p ComboboxEmptyProps) templ.Attributes {
	// The no-results message lives inside the listbox, so it carries role="option"
	// (an allowed listbox child) + aria-disabled rather than role="status" (which a
	// listbox may not contain). It is data-slot="empty" so the controller never
	// treats it as a selectable option, and hidden until Open.
	owned := templ.Attributes{
		"data-slot":     "empty",
		"role":          "option",
		"aria-disabled": "true",
	}
	if !p.Open {
		owned["hidden"] = true
	}
	return ownedAttrs(p.Base, owned)
}

File: ui/goth/assets/src/js/controllers/combobox.js
// gothCombobox controller (GOTH-1.4; overlay mechanics GOTH-4.1; command/combobox
// primitives GOTH-5.2).
//
// Backs the Combobox (P50) and Command (P51) primitives from one controller with no
// new §8 name. Two shapes:
//   - popup (Combobox): the listbox opens on focus/typing and dismisses on
//     Escape/outside press through the shared overlay mechanics;
//   - inline (Command, data-inline): the listbox is always visible, no overlay.
//
// In both, the input keeps DOM focus and an active option is tracked with
// aria-activedescendant (APG editable-combobox pattern): the arrow keys move it,
// Enter activates it (a native submit/link — the server owns the round-trip),
// Home/End jump to the ends. Filtering is client-side (hide non-matching options +
// toggle the empty state) or server-owned (data-filter="server": the server
// replaces the options via HTMX and renders its own empty state). Dispatches the
// documented goth:open / goth:close / goth:select events (README §8). The server
// still owns the submitted value.

import { createDismisser } from "../mechanics/dismiss.js";

export default function gothCombobox() {
  return {
    open: false,
    _root: null,
    _input: null,
    _listbox: null,
    _empty: null,
    _inline: false,
    _serverFilter: false,
    _dismisser: null,
    _active: -1,

    init() {
      // Cache the root at init; a method invoked from a descendant's x-on handler
      // sees $el bound to that descendant (the GOTH-4.1 descendant-$el fix).
      this._root = this.$el;
      this._input = this._root.querySelector('[data-slot="input"]');
      this._listbox = this._root.querySelector('[data-slot="listbox"]');
      this._empty = this._root.querySelector('[data-slot="empty"]');
      this._inline = this._root.hasAttribute("data-inline");
      this._serverFilter = this._root.getAttribute("data-filter") === "server";
      // Enhanced: keyboard is activedescendant on the input, so options leave the
      // tab order (one tab stop). No-JS keeps the buttons natively focusable.
      this._demoteOptions();
      if (this._inline) {
        this.open = true;
        this._root.setAttribute("data-state", "open");
        if (this._input) this._input.setAttribute("aria-expanded", "true");
      }
      // Re-index options after an HTMX swap replaces the server-owned option list.
      if (this._listbox) {
        this._listbox.addEventListener("htmx:afterSwap", () => this._afterSwap());
      }
    },

    _options() {
      if (!this._root) return [];
      return Array.from(this._root.querySelectorAll('[data-slot="option"]'));
    },

    _visibleOptions() {
      return this._options().filter(
        (o) =>
          !o.hidden &&
          !o.hasAttribute("disabled") &&
          o.getAttribute("aria-disabled") !== "true",
      );
    },

    _demoteOptions() {
      this._options().forEach((o) => o.setAttribute("tabindex", "-1"));
    },

    _afterSwap() {
      this._demoteOptions();
      this._setActive(-1);
      this._syncEmpty();
    },

    _ensureID(opt, i) {
      if (!opt.id) {
        opt.id = (this._root.id || "goth-combobox") + "-opt-" + i;
      }
      return opt.id;
    },

    onFocus() {
      if (!this._inline) this.show();
    },

    onInput() {
      if (!this._serverFilter) this._clientFilter();
      if (!this._inline && !this.open) this.show();
      this._syncEmpty();
    },

    _clientFilter() {
      const q = (this._input ? this._input.value : "").trim().toLowerCase();
      this._options().forEach((opt) => {
        const label = (opt.textContent || "").trim().toLowerCase();
        opt.hidden = q !== "" && !label.includes(q);
      });
      this._setActive(-1);
    },

    _syncEmpty() {
      // In server-filter mode the server renders the empty state; only toggle it
      // for client filtering.
      if (!this._empty || this._serverFilter) return;
      this._empty.hidden = this._visibleOptions().length > 0;
    },

    show() {
      if (this.open) return;
      this.open = true;
      this._root.setAttribute("data-state", "open");
      if (this._input) this._input.setAttribute("aria-expanded", "true");
      // Escape refocuses the input (it never lost focus anyway); an outside press
      // just closes.
      this._dismisser = createDismisser(this._root, (reason) =>
        this.hide(reason === "escape"),
      );
      this.$nextTick(() => {
        if (this._dismisser) this._dismisser.activate();
        this.$dispatch("goth:open");
      });
    },

    hide(refocus = true) {
      if (!this.open || this._inline) return;
      this.open = false;
      this._root.setAttribute("data-state", "closed");
      if (this._input) {
        this._input.setAttribute("aria-expanded", "false");
        this._input.removeAttribute("aria-activedescendant");
        if (refocus) this._input.focus();
      }
      this._setActive(-1);
      if (this._dismisser) {
        this._dismisser.deactivate();
        this._dismisser = null;
      }
      this.$dispatch("goth:close");
    },

    _setActive(target) {
      this._options().forEach((o) => o.removeAttribute("data-active"));
      const opts = this._visibleOptions();
      if (opts.length === 0 || target < 0) {
        this._active = -1;
        if (this._input) this._input.removeAttribute("aria-activedescendant");
        return;
      }
      this._active = ((target % opts.length) + opts.length) % opts.length;
      const opt = opts[this._active];
      const id = this._ensureID(opt, this._active);
      opt.setAttribute("data-active", "true");
      if (this._input) this._input.setAttribute("aria-activedescendant", id);
      if (typeof opt.scrollIntoView === "function") {
        opt.scrollIntoView({ block: "nearest" });
      }
    },

    _activeOption() {
      const opts = this._visibleOptions();
      if (this._active < 0 || this._active >= opts.length) return null;
      return opts[this._active];
    },

    onKeydown(event) {
      switch (event.key) {
        case "ArrowDown":
          event.preventDefault();
          if (!this.open && !this._inline) {
            this.show();
            this.$nextTick(() => this._setActive(0));
          } else {
            this._setActive(this._active + 1);
          }
          break;
        case "ArrowUp":
          event.preventDefault();
          if (!this.open && !this._inline) {
            this.show();
            this.$nextTick(() => this._setActive(this._visibleOptions().length - 1));
          } else {
            this._setActive(this._active - 1);
          }
          break;
        case "Home":
          if (this.open || this._inline) {
            event.preventDefault();
            this._setActive(0);
          }
          break;
        case "End":
          if (this.open || this._inline) {
            event.preventDefault();
            this._setActive(this._visibleOptions().length - 1);
          }
          break;
        case "Enter": {
          const opt = this._activeOption();
          if (opt) {
            event.preventDefault();
            opt.click();
          }
          break;
        }
        case "Escape":
          if (this.open && !this._inline) {
            event.preventDefault();
            this.hide(true);
          } else if (this._inline) {
            this.$dispatch("goth:close");
          }
          break;
        case "Tab":
          if (this.open && !this._inline) this.hide(false);
          break;
      }
    },

    select(event) {
      // The option is a native submit button or link: let the browser (no-JS) or
      // HTMX perform the server round-trip. We only report the selection.
      this.$dispatch("goth:select", {
        value: event.currentTarget.getAttribute("data-value"),
      });
    },
  };
}
